### PR TITLE
Add tests for expression variables in nameof()

### LINKED ISSUE #15604: Add tests for declarations appearing in a nameof argument
Since it is *possible* to write a general expression inside of a `nameof()`, such as an expression `(int x, int y)`, `(M(out var x), var (y, z) = e)` etc which are tuple expressions containing declaration expressions, we may want the SemanticModel to work for these declaration expressions and their types, etc. I do not believe we have SemanticModel support for such expressions today, or tests for these scenarios.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #15604 

This is a test-only change, and so not subject to bar checks and shiproom approval.

@dotnet/roslyn-compiler May I please have a review of this test-only change?

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs [hunk 1]
```diff
@@ -2729,5 +2729,57 @@ public void Deconstruct(out dynamic x, out dynamic y)
...
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+            var designations = tree.GetCompilationUnitRoot().DescendantNodes().OfType<SingleVariableDesignationSyntax>().ToArray();
+            Assert.Equal(2, designations.Count());
+            var refs = tree.GetCompilationUnitRoot().DescendantNodes().OfType<IdentifierNameSyntax>();
+
+            var x1 = model.GetDeclaredSymbol(designations[0]);
+            Assert.Equal("x1", x1.Name);
+            Assert.Equal("System.Int32", ((LocalSymbol)x1).Type.ToTestDisplayString());
+            Assert.Equal(x1, model.GetSymbolInfo(refs.Where(r => r.Identifier.ValueText == "x1").Single()).Symbol);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs]: >Equal [](start = 19, length = 5)

Same?

File: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs [hunk 2]
```diff
@@ -2729,5 +2729,57 @@ public void Deconstruct(out dynamic x, out dynamic y)
...
+
+            var x2 = model.GetDeclaredSymbol(designations[1]);
+            Assert.Equal("x2", x2.Name);
+            Assert.Equal("System.Int32", ((LocalSymbol)x2).Type.ToTestDisplayString());
+            Assert.Equal(x2, model.GetSymbolInfo(refs.Where(r => r.Identifier.ValueText == "x2").Single()).Symbol);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs]: >Equal [](start = 19, length = 5)

Same?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_Scope.cs [hunk 3]
```diff
@@ -11481,5 +11481,45 @@ static bool Dummy(params object[] data)
             var zRef = GetReferences(tree, "z1").Single();
             Assert.Equal("System.Boolean", compilation.GetSemanticModel(tree).GetTypeInfo(zRef).Type.ToTestDisplayString());
         }
+
+
+        [Fact]
+        public void DeclarationInsideNameof()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_Scope.cs]: >public void DeclarationInsideNameof() [](start = 8, length = 37)

There is an extra empty line before this method.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_Scope.cs [hunk 4]
```diff
@@ -11481,5 +11481,45 @@ static bool Dummy(params object[] data)
...
+                //         string s1 = x1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "x1").WithArguments("int", "string").WithLocation(7, 21),
+                // (7,21): error CS0165: Use of unassigned local variable 'x1'
+                //         string s1 = x1;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x1").WithArguments("x1").WithLocation(7, 21)
+                );
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+            var designations = tree.GetCompilationUnitRoot().DescendantNodes().OfType<SingleVariableDesignationSyntax>().ToArray();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_Scope.cs]: >tree.GetCompilationUnitRoot().DescendantNodes().OfType<SingleVariableDesignationSyntax>().ToArray(); [](start = 31, length = 100)

Can we use GetPatternDeclarations helper?

File: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs]: I mean Assert.Same?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_Scope.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_Scope.cs]: By not restricting the identifier as `GetPatternDeclarations` does, I am asserting that there are no other designators in this program.

**Pull request merged**